Make README more useful (#39)
ARMmbed/mbed-os-example-crash-reporting: README.md

![](./resources/official_armmbed_example_badge.png)
# Crash Reporting Mbed OS example

> This guide reviews the steps required to get Mbed OS crash-reporting on a Mbed OS enabled platform.

Reviewer: ```suggestion
This guide reviews the steps required to get Mbed OS crash reporting on a Mbed OS enabled platform.
```

> You can find complete information about crash reporting architecture, high-level design and APIs [here](https://github.com/ARMmbed/mbed-os/blob/master/docs/design-documents/platform/crash-reporting/crash_reporting.md#deprecations)

Reviewer: ```suggestion
You can find complete information about crash reporting architecture, high-level design and APIs [here](https://github.com/ARMmbed/mbed-os/blob/master/docs/design-documents/platform/crash-reporting/crash_reporting.md).
```

Reviewer: More importantly than the high level design is the link to the APIs in the documentation: https://os.mbed.com/docs/mbed-os/v5.15/apis/error-handling.html#crash-reporting-and-auto-reboot

You can build this project with all supported [Mbed OS build tools](https://os.mbed.com/docs/mbed-os/latest/tools/index.html). However, this example project specifically refers to the command-line interface tool [Arm Mbed CLI](https://github.com/ARMmbed/mbed-cli#installing-mbed-cli).

## Application functionality

> This application does the following:

Reviewer: I think it would be useful to describe the functionality in two passes.
In the first pass, a fault is generated.
In the second pass, we retrieve the error context of the reboot.

> - The `main()` function purposely causes an unaligned memory access exception, which causes an auto-reboot of the system.

Reviewer: I would not emphasize `main()` in this section to only leave the APIs emphasized.

> - After rebooting, the system detects that the reboot was due to a fatal error and prints out information about the previous error context using `mbed_error_reboot_callback()` and clears that information from RAM. This happens before executing `main()` again.

Reviewer: ```suggestion
- After rebooting, during Mbed OS initialization, the system checks the Crash-data-RAM region, detects that the reboot was due to a fatal error and invokes the `mbed_error_reboot_callback()` callback function with a pointer to the error context structure stored in Crash-data-RAM. In this example, the callback function was overriden to print information about the error context and then clear that information from RAM.
```

> - When `main()` is executed again, it is notified that a reboot occurred and thus proceeds with printing the fault context after retrieving it with `mbed_get_reboot_fault_context()`.

Reviewer: Be more precise. How is main notified that a reboot occurred?

### Note:

> There are two ways to copy the error context:

Reviewer: ```suggestion
There are two ways to retrieve the error context:
```

1. Copy the error context from within `mbed_error_reboot_callback()` using the argument it receives.

> 2. Use `mbed_get_reboot_error_info`.

Reviewer 2: ```suggestion
2. Use `mbed_get_reboot_error_info()`.
```
For consistency.

> 2. Use `mbed_get_reboot_error_info()`.

Reviewer: Rework a bit the above two sentences and give more information on the differences between the two. Alternatively you could show usage of `mbed_get_reboot_error_info()` in main().

Use the information below:
`mbed_get_reboot_error_info()`: The below API can be called by application to retrieve the error context captured in the Crash-Report RAM. The error context is copied into the location pointed by error_info. Note that the caller should allocate the memory for this location. The function should return MBED_ERROR_NOT_FOUND if there is no error context currently stored.

`void mbed_error_reboot_callback(mbed_error_ctx *error_context);`: This callback is invoked before the system starts executing the application main(). The implementation of callback should be aware any resource limitations or availability. Also, the callback is invoked only when there is a new error.

![](./resources/official_armmbed_example_badge.png)
# Crash Reporting Mbed OS example

> You can find complete information about crash reporting architecture, high-level design and APIs [here](https://github.com/ARMmbed/mbed-os/blob/master/docs/design-documents/platform/crash-reporting/crash_reporting.md#deprecations)

Reviewer 2: Maybe swap line 4 and line 6

This guide reviews the steps required to get Mbed OS crash-reporting on a Mbed OS enabled platform.

You can build this project with all supported [Mbed OS build tools](https://os.mbed.com/docs/mbed-os/latest/tools/index.html). However, this example project specifically refers to the command-line interface tool [Arm Mbed CLI](https://github.com/ARMmbed/mbed-cli#installing-mbed-cli).

## Application functionality

This application does the following:

> - The `main()` function injects an unaligned memory access exception(Mbed OS fault exception), which causes an auto-reboot(warm-reset) of the system. 

Reviewer 2: ```suggestion
- The `main()` function purposely causes an unaligned memory access exception, which causes an auto-reboot of the system. 
```

> - During the reboot, the system will detect that the reboot and it calls the `mbed_error_reboot_callback()` with the error context retrieved from the RAM. 

Reviewer 2: ```suggestion
- After rebooting, the system detects that the reboot was due to a fatal error and prints out information about the previous error context using `mbed_error_reboot_callback()` and clears that information from RAM. This happens before executing `main()` again.
```

- The callback function does the following:

   - Prints error context information such as status, value, address, reboot count and CRC

   - Copy the error_status, which helps to identify the type of exception

>    - Enable `reboot_error_happened` global flag and reset the error context

Reviewer 2: ```suggestion
```
Lines 17 to 23 to be deleted.

> - This time, `main()` detects that a reboot using `reboot_error_happened` flag and retrieves the fault context information from RAM using mbed_get_reboot_fault_context.

Reviewer 2: ```suggestion
- When `main()` is executed again, it is notified that a reboot occurred and thus proceeds with printing the fault context after retrieving it with `mbed_get_reboot_fault_context()`.
```

### Note:

> There are two ways to copy the error context in the `mbed_error_reboot_callback()`

Reviewer 2: ```suggestion
There are two ways to copy the error context:
```

> 1. Use `memcpy` with callback argument `error_context` to global memory of type `mbed_error_ctx`

Reviewer 2: ```suggestion
1. Copy the error context from within `mbed_error_reboot_callback()` using the argument it receives.
```

> 2. Pass the global memory as a argument to `mbed_get_reboot_error_info` to retrieve. 

Reviewer 2: ```suggestion
2. Use `mbed_get_reboot_error_info()`. 
```

![](./resources/official_armmbed_example_badge.png)
# Crash reporting Mbed OS example

This example demonstrates how Mbed OS crash reporting works on an Mbed OS Enabled platform.

You can find more information about the crash reporting APIs [in the documetnation](https://os.mbed.com/docs/mbed-os/latest/apis/error-handling.html#crash-reporting-and-auto-reboot).

You can build this project with all supported [Mbed OS build tools](https://os.mbed.com/docs/mbed-os/latest/tools/index.html). However, this example project specifically refers to the command-line interface tool [Arm Mbed CLI](https://github.com/ARMmbed/mbed-cli#installing-mbed-cli).

1. Install Mbed CLI.
1. Clone this repository on your system.
1. Change the current directory to where the project was cloned.

## Application functionality

> This example demonstrates the Mbed OS crash reporting feature and the APIs associated with it. In the first pass, the application generates a fault that causes the system to reboot. In the second pass, the application detects that the reboot was caused by an error and prints the fault context information in the console.

Reviewer 3: Would it make sense to say "first part" or "first stage" instead of "first pass"?

Reviewer: We can say first run instead.

First pass:

1. During Mbed OS initialization, the system checks the Crash-data-RAM region and detects there was no fault.

> 1. `main()` is called.

Reviewer 3: What calls `main()`? You, the user?

>    1. An exception is generated.

Reviewer 3: What generates the exception?

   1. The system captures the error-context in a special location in RAM.
1. The system auto-reboots (warm-resets).

Second pass:

1. During Mbed OS initialization, the system:
   1. Checks the Crash-data-RAM region,.
   1. Detects that the reboot was due to a fatal error.
   1. Calls `mbed_error_reboot_callback()` with a pointer to the error context stored in RAM.
   
   In this example:

>       1. The callback function (WEAK function) is overridden.

Reviewer 3: What overrides the function?

>       1. A global variable `reboot_error_happened` is set to record that reboot was due to an error.

Reviewer 3: What sets the variable? You, the user?

      1. The function prints information about the error context in the console and resets the saved context captured by the system in RAM using `mbed_reset_reboot_error_info()`.
      

> 1. `main()` is called.

Reviewer 3: What calls `main()`?

++ MbedOS Error Info ++
Error Status: 0x80FF013D Code: 317 Module: 255
Error Message: Fault exception
Location: 0x178C
Error Value: 0x1FFF0400
Current Thread: main Id: 0x20000E68 Entry: 0x319B StackSize: 0x1000 StackMem: 0x20001E60 SP: 0x20002E50 
For more info, visit: https://mbed.com/s/error?error=0x80FF013D&tgt=K64F
-- MbedOS Error Info --

= System will be rebooted due to a fatal error =
= Reboot count(=1) reached maximum, system will halt after rebooting =
 Error:
Status: 0x80FF013D
Value: 0x1FFF0400
Address: 0x178C
Reboot count: 0x1
CRC: 0x958A47DB
This is the crash reporting Mbed OS example
0: 0xDE0C
1: 0x0
2: 0xE000ED00
3: 0xAAA3
4: 0x0
5: 0x0
6: 0x0
7: 0x0
8: 0x0
9: 0x0
10: 0x0
11: 0x0
12: 0x1FFF052C
13: 0x20002E50
14: 0x1817
15: 0x178C
```
## Crash-report platform configuration

> The crash report feature can be enabled by `platform.crash-capture-enabled: true` configuration in `mbed_app.json` as below:

Reviewer: Say that this is enabled on some targets by default. I assume that if it is not enabled then the loader/sactter file should be modified, in which case you could provide a link to explain what needs to be done.

++ MbedOS Error Info ++
Error Status: 0x80FF013D Code: 317 Module: 255
Error Message: Fault exception
Location: 0x178C
Error Value: 0x1FFF0400
Current Thread: main Id: 0x20000E68 Entry: 0x319B StackSize: 0x1000 StackMem: 0x20001E60 SP: 0x20002E50 
For more info, visit: https://mbed.com/s/error?error=0x80FF013D&tgt=K64F
-- MbedOS Error Info --

= System will be rebooted due to a fatal error =
= Reboot count(=1) reached maximum, system will halt after rebooting =
 Error:
Status: 0x80FF013D
Value: 0x1FFF0400
Address: 0x178C
Reboot count: 0x1
CRC: 0x958A47DB
This is the crash reporting Mbed OS example
0: 0xDE0C
1: 0x0
2: 0xE000ED00
3: 0xAAA3
4: 0x0
5: 0x0
6: 0x0
7: 0x0
8: 0x0
9: 0x0
10: 0x0
11: 0x0
12: 0x1FFF052C
13: 0x20002E50
14: 0x1817
15: 0x178C
```
## Crash-report platform configuration

The crash report feature can be enabled by `platform.crash-capture-enabled: true` configuration in `mbed_app.json` as below:
```
    "target_overrides": {
        "*": {            
            "platform.crash-capture-enabled": true            
        }
```

>  and disable with `platform.crash-capture-enabled: false`

Reviewer 2: ```suggestion
```
It goes without saying.
